cdap/resource_streaming_program_run.go

```diff
@@ -0,0 +1,335 @@
+	// cast map[string]interface{} to map[string]string
+	for k, val := range d.Get("runtime_arguments").(map[string]interface{}) {
+		argsObj[k] = val.(string)
+	}
+
+	runId, _ := uuid.NewRandom()
+	// This runtime arg will be unused by the pipeline but will allow the provider to associate a run with this resource.
+	argsObj[FAUX_RUN_ID] = runId.String()
+
+	b, err := json.Marshal(argsObj)
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest(http.MethodPost, startAddr, bytes.NewReader(b))
+	if err != nil {
+		return err
+	}
+
+	if _, err := httpCall(config.httpClient, req); err != nil {
+		return err
+	}
+
+	// Poll until actually reaches RUNNING state.
+	return resource.Retry(d.Timeout(schema.TimeoutCreate), func() *resource.RetryError {
+		isRunning, err := isFauxRunIdRunningYet(config, runsAddr, runId.String())
+		if err != nil {
+			return resource.RetryableError(err)
+		}
+		if isRunning {
+			d.SetId(runId.String())
+			return nil
+		}
+		time.Sleep(10 * time.Second) // avoid spamming retries
+		return resource.RetryableError(fmt.Errorf("still waiting for program run with faux run id: %s which is in an initializing state", runId.String()))
+	})
+}
+
+//TODO ?
+func resourceStreamingProgramRunRead(d *schema.ResourceData, m interface{}) error {
+	return nil
+}
+
+type RuntimeArgs struct {
+	FauxRunId string `json:"__FAUX_RUN_ID__"`
+}
+
+type RuntimeProperties struct {
+	RuntimeArgs json.RawMessage `json:"runtimeArgs"`
+}
+
+// These are the only keys we need
+type Run struct {
+	RunId      string `json:"runid"`
+	Status     string `json:"status"`
+	Properties RuntimeProperties
+}
+
+// Checks if there is a running run for the terraform faux run id
+// raises error if the program is not in an initializing state (e.g. it failed or was killed in the ui)
+func isFauxRunIdRunningYet(config *Config, runsAddr string, runId string) (bool, error) {
+	s, err := getProgramRunStatusByFauxId(config, runsAddr, runId)
+	if err != nil {
+		return false, err
+	}
+
+	if s == "RUNNING" {
+		return true, nil
+	}
+	if !INITIALIZING_PROGRAM_RUN_STATUSES[s] {
+		return false, fmt.Errorf("Program not Running or Initializing, in state: %s", s)
+	}
+	return false, nil
+
+}
+
+func getProgramRunStatusByFauxId(config *Config, runsAddr string, runId string) (s string, err error) {
+	r, err := getRunByFauxId(config, runsAddr, runId)
+	if err != nil {
+		return
+	}
+
+	return r.Status, nil
+}
+
+// TODO optimization: we call this function often when we could probably get the real runid once and cache it.
+// This would avoid redoing this loop everytime to get the same result. probably inconsequential unless there are many runs of this program
+func getRunByFauxId(config *Config, runsAddr string, runId string) (r Run, err error) {
+	req, err := http.NewRequest(http.MethodGet, runsAddr, nil)
+	if err != nil {
+		return
+	}
+
+	b, err := httpCall(config.httpClient, req)
+	if err != nil {
+		return
+	}
+
+	var runs []Run
+	var args RuntimeArgs
```

Comment:
Do args need to be defined here? Can you moveit into the loop?

Reply:
we can move it in the loop.
Is there value in doing that? 
I assumed it might be better to not re declare it in each iteration.

Reply:
Prefer to define vars close to the point they are used and also define them in as small a scope as possible